pkg/buses/natss/bus.go

```diff
@@ -0,0 +1,178 @@
+			return nil
+		},
+	}
+	setup(bus)
+	return bus, nil
+}
+
+func NewNatssBusDispatcher(ref buses.BusReference, setup SetupNatssBus) (*NatssBus, error) {
+	bus := &NatssBus{
+		ref:         ref,
+		subscribers: make(map[string]*stan.Subscription),
+	}
+	bus.dispEventHandler = buses.EventHandlerFuncs{
+		ReceiveMessageFunc: func(channel buses.ChannelReference, message *buses.Message) error {
+			bus.logger.Infof("Recieved message from %q channel", channel.String())
+			if err := stanutil.Publish(natsConn, channel.Name, &message.Payload, bus.logger); err != nil {
+				bus.logger.Errorf("Error during publish: %+v", err)
+				return err
+			}
+			bus.logger.Infof("Published [%s] : '%s'", channel.String(), message)
+			return nil
+		},
+		SubscribeFunc:   bus.subscribe,
+		UnsubscribeFunc: bus.unsubscribe,
+	}
+	setup(bus)
+	return bus, nil
+}
+
+func (b *NatssBus) Run(threadness int, stopCh <-chan struct{}, clientId string) {
+	b.logger.Infof("try to connect to NATSS from %q", clientId)
+	var err error
+	for i := 0; i < 60; i++ {
+		if natsConn, err = stanutil.Connect("knative-nats-streaming", clientId, b.natssUrl, b.logger); err != nil {
+			b.logger.Errorf(" Create new connection failed: %+v", err)
+			time.Sleep(1 * time.Second)
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		b.logger.Errorf(" Create new connection failed: %+v", err)
+		return
+	}
+	b.logger.Info("connection to NATSS established, natsConn=%+v", natsConn)
+
+	if b.dispatcher != nil {
+		b.dispatcher.Run(threadness, stopCh)
+	}
+	if b.provisioner != nil {
+		b.provisioner.Run(threadness, stopCh)
+	}
+}
+
+func SetNewBusProvisioner(opts *buses.BusOpts) SetupNatssBus {
+	return func(b *NatssBus) error {
+		b.natssUrl = NatssUrl
+		b.provisioner = buses.NewBusProvisioner(b.ref, b.provEventHandler, opts)
+		b.logger = opts.Logger
+		return nil
+	}
+}
+
+func SetNewBusDispatcher(opts *buses.BusOpts) SetupNatssBus {
+	return func(b *NatssBus) error {
+		b.natssUrl = NatssUrl
+		b.dispatcher = buses.NewBusDispatcher(b.ref, b.dispEventHandler, opts)
+		b.logger = opts.Logger
+		return nil
+	}
+}
+
+func (bus *NatssBus) subscribe(channel buses.ChannelReference, subscription buses.SubscriptionReference, parameters buses.ResolvedParameters) error {
+	bus.logger.Infof("Subscribe %q to %q channel", subscription.Name, channel.Name)
+
+	mcb := func(msg *stan.Msg) {
+		bus.logger.Infof("NATSS message received: %+v", msg)
+		message := buses.Message{
+			Headers: map[string]string{},
+			Payload: []byte(msg.Data),
+		}
+		if err := bus.dispatcher.DispatchMessage(subscription, &message); err != nil {
+			bus.logger.Warnf("Failed to dispatch message: %v", err)
+			return
+		}
+		if err := msg.Ack(); err != nil {
+			bus.logger.Warnf("Failed to acknowledge message: %v", err)
+		}
+	}
+	// subscribe to a NATSS subject
+	if natsStreamingSub, err := (*natsConn).Subscribe(channel.Name, mcb, stan.DurableName(subscription.Name), stan.SetManualAckMode(), stan.AckWait(1*time.Minute)); err != nil {
```

Comment:
are durable name sufficient to prevent duplicate delivery for multiple subscriptions for the same topic. If so, the name will need to be much more robust that simply the subscription name as the same subscription name in a different namespace will collide.

The name should also be robust against the same resource existing in duplicate k8s clusters. This is something that led to a bit of confusion with the Cloud Pub/Sub bus.

Reply:
Also ackWaitTimeout should be a const and moved to the top of the file.